crypto/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,267 @@
+use crate::hash::traits::IsCryptoHash;
+use lambdaworks_math::field::{element::FieldElement, traits::IsField};
+use std::{cell::RefCell, rc::Rc};
+
+pub struct MerkleTree<F: IsField, H: IsCryptoHash<F>> {
+    leafs: Vec<TreeNode<F>>,
+    root: TreeNode<F>,
+    hasher: H,
+}
+
+impl<F: IsField, H: IsCryptoHash<F> + Clone> MerkleTree<F, H> {
+    pub fn build(values: &[FieldElement<F>], hasher: H) -> MerkleTree<F, H> {
+        let mut leafs: Vec<TreeNode<F>> = hash_leafs(values, &hasher);
+        let mut level: Vec<TreeNode<F>> = hash_level(&mut leafs, &hasher);
+
+        while level.len() > 1 {
+            level = hash_level(&mut level, &hasher);
+        }
+
+        MerkleTree {
+            leafs,
+            root: level[0].clone(),
+            hasher,
+        }
+    }
+
+    pub fn get_root_hash(&self) -> FieldElement<F> {
+        self.root.borrow().hash.clone()
+    }
+
+    pub fn get_proof(&self, value: FieldElement<F>) -> Option<Proof<F, H>> {
+        let hashed_value = self.hasher.hash_one(value.clone());
+
+        if let Some(leaf) = self
+            .leafs
+            .iter()
+            .find(|node| node.borrow().hash == hashed_value)
+        {
+            let mut merkle_path: Vec<TreeNode<F>> = Vec::new();
+            merkle_path = build_merkle_path(leaf.clone(), &mut merkle_path).to_vec();
```

Comment:
These lines could be collapsed like this (if I'm not mistaken)
```suggestion
            let merkle_path = build_merkle_path(leaf.clone(), &mut Vec::new()).to_vec();
```